% Created 2014-06-21 Sat 16:01
\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{fixltx2e}
\usepackage{graphicx}
\usepackage{longtable}
\usepackage{float}
\usepackage{wrapfig}
\usepackage{soul}
\usepackage{textcomp}
\usepackage{marvosym}
\usepackage{wasysym}
\usepackage{latexsym}
\usepackage{amssymb}
\usepackage{hyperref}
\tolerance=1000
\author{Chen Alvin}
\date{\today}
\title{MSLTEasy Help Content}
\hypersetup{
  pdfkeywords={},
  pdfsubject={},
  pdfcreator={Generated by Org mode 7.9.3f in Emacs 24.3.50.2.}}
\begin{document}

\maketitle
\tableofcontents

\section[Introduction]{Introduction}
\label{sec-1}
Efficient methods for estimating genetic diversity among the microorganisms are essential for understanding evolutionary history, geographic distribution, pathogenicity and virulence. In the past decades, numerous methods have been developed for typing of bacteria and fungi. Multilocus sequence typing (MLST) based DNA sequencing results, which can be easily archived and shared among different laboratories. MLST is one of the most reliable and informative method for molecular genotyping, and it has been adopted in many bacterial and fungal studies. 

MLSTEasy was designed for next generation sequencing technology (PacBio CCS or Roche 454 platform) based MSLT methods, which were previously described \footnote{DEFINITION NOT FOUND.}\textsuperscript{,}\,\footnote{DEFINITION NOT FOUND.}. MLST-Easy, can automatically identify the barcodes and primers used in the PCR reaction, corrects sequencing errors, generates the MLST profile for each isolate, predicts the potential heterozygous locus, and outputs different alleles. 

\section[System requirements]{System requirements}
\label{sec-2}
MLSTEasy was written in Python, version 2.7.6. The graphic user interface (GUI) was created by PyQt4 (\url{http://www.riverbankcomputing.com/software/pyqt/download}) and Qt Designer (\url{http://qt-project.org/doc/qt-4.8/designer-manual.html}). Mac version were tested under Mac OS X 10.9, and Windows version were tested under Windows Vista and Windows 7. The software runs on IBM-compatible PC under 32/64-bit Windows, and Mac OS X 10.6+. The minimum hardware requirements for the program are:
  a processor based on the Intel Pentium 4/AMD Athlon
  200 MB of RAM memory
  hard drive with more than 200 MB available space
  Windows XP or later version, Mac OS X 10.6+ with MUSCLE installed (\url{http://www.drive5.com/muscle/downloads.htm})

The recommended hardware requirements for the program are:
  a multiple core processor based on the Intel Core 2 Due/AMD Athlon II (or higher)
  4 GB of RAM memory
  hard drive with more than 1 GB available space
  Windows XP or later version, Mac OS X 10.6+ with MUSCLE installed (\url{http://www.drive5.com/muscle/downloads.htm})
\section[Create a new project]{Create a new project}
\label{sec-3}
A new project can be created by click the new button on tool bar \includegraphics[width=.9\linewidth]{./Figures/New.png}
\subsection[Sequencing Files (FASTA format; FASTQ format)]{Sequencing Files (FASTA format; FASTQ format)}
\label{sec-3-1}
\subsection[Barcode File]{Barcode File}
\label{sec-3-2}
\subsection[Primer File]{Primer File}
\label{sec-3-3}
\subsection[Parameters]{Parameters}
\label{sec-3-4}

\section[Run project]{Run project}
\label{sec-4}
\subsection[Barcode and primer identification]{Barcode and primer identification}
\label{sec-4-1}
\subsection[Generate consensus sequences]{Generate consensus sequences}
\label{sec-4-2}
\subsection[Dump unmapped reads]{Dump unmapped reads}
\label{sec-4-3}
\subsection[Search for heterozygous loci]{Search for heterozygous loci}
\label{sec-4-4}

\section[Open Project]{Open Project}
\label{sec-5}
\section[Merge Project]{Merge Project}
\label{sec-6}
\section[Result]{Result}
\label{sec-7}
% Generated by Org mode 7.9.3f in Emacs 24.3.50.2.
\end{document}